jaxpm/pm.py

```diff
@@ -1,48 +1,80 @@
 import jax
 import jax.numpy as jnp
 import jax_cosmo as jc
+from jax_cosmo import Cosmology
 
-from jaxpm.growth import dGfa, growth_factor, growth_rate
-from jaxpm.kernels import (PGD_kernel, fftk, gradient_kernel, laplace_kernel,
-                           longrange_kernel)
+from jaxpm.growth import growth_factor, growth_rate, dGfa, growth_factor_second, growth_rate_second, dGf2a
+from jaxpm.kernels import PGD_kernel, fftk, gradient_kernel, invlaplace_kernel, longrange_kernel
 from jaxpm.painting import cic_paint, cic_read
 
 
-def pm_forces(positions, mesh_shape=None, delta=None, r_split=0):
+
+def pm_forces(positions, mesh_shape, delta=None, r_split=0):
     """
     Computes gravitational forces on particles using a PM scheme
     """
-    if mesh_shape is None:
-        mesh_shape = delta.shape
-    kvec = fftk(mesh_shape)
-
     if delta is None:
         delta_k = jnp.fft.rfftn(cic_paint(jnp.zeros(mesh_shape), positions))
-    else:
+    elif jnp.isrealobj(delta):
         delta_k = jnp.fft.rfftn(delta)
+    else:
+        delta_k = delta
 
     # Computes gravitational potential
-    pot_k = delta_k * laplace_kernel(kvec) * longrange_kernel(kvec,
-                                                              r_split=r_split)
+    kvec = fftk(mesh_shape)
+    pot_k = delta_k * invlaplace_kernel(kvec) * longrange_kernel(kvec, r_split=r_split)
     # Computes gravitational forces
-    return jnp.stack([
-        cic_read(jnp.fft.irfftn(gradient_kernel(kvec, i) * pot_k), positions)
-        for i in range(3)
-    ],
-                     axis=-1)
+    return jnp.stack([cic_read(jnp.fft.irfftn(- gradient_kernel(kvec, i) * pot_k), positions) 
+                      for i in range(3)], axis=-1)
 
 
-def lpt(cosmo, initial_conditions, positions, a):
+def lpt(cosmo:Cosmology, init_mesh, positions, a, order=1):
     """
-    Computes first order LPT displacement
+    Computes first and second order LPT displacement and momentum, 
+    e.g. Eq. 2 and 3 [Jenkins2010](https://arxiv.org/pdf/0910.0258)
     """
-    initial_force = pm_forces(positions, delta=initial_conditions)
     a = jnp.atleast_1d(a)
-    dx = growth_factor(cosmo, a) * initial_force
-    p = a**2 * growth_rate(cosmo, a) * jnp.sqrt(jc.background.Esqr(cosmo,
-                                                                   a)) * dx
-    f = a**2 * jnp.sqrt(jc.background.Esqr(cosmo, a)) * dGfa(cosmo,
-                                                             a) * initial_force
+    E = jnp.sqrt(jc.background.Esqr(cosmo, a)) 
+    delta_k = jnp.fft.rfftn(init_mesh) # TODO: pass the modes directly to save one or two fft?
+    mesh_shape = init_mesh.shape
+
+    init_force = pm_forces(positions, mesh_shape, delta=delta_k)
```

Comment:
@EiffL 
Should we pass kvec as an argument to pm_forces?
To avoid creating it twice in case of distribution and LPT2?